File: Python/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode()
        cur = res
        overflow = 0

        while l1 or l2 or overflow != 0:
            temp = overflow
            if l1: temp += l1.val
            if l2: temp += l2.val

            cur.val = temp % 10
            overflow = temp // 10

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

            if not (l1 or l2 or overflow != 0): break

            next = ListNode()
            cur.next = next
            cur = next

        return res

File: C/2. Add Two Numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    int temp = 0;
    struct ListNode* t1 = l1;
    struct ListNode* t2 = l2;

    while (t1) {
        if (t1 && t2 && (!t1->next && t2->next)) {
            t1->next = t2->next;
            t2->next = NULL;
        }

        if (t2) {
            temp = temp + t1->val + t2->val;
            t1->val = temp % 10;
            temp /= 10;
            t1 = t1->next; t2 = t2->next;
        }
        else {
            temp = temp + t1->val;
            t1->val = temp % 10;
            temp /= 10;

            if (temp == 0) break;
            else if (!t1->next) {
                struct ListNode* new_node = (struct ListNode*)malloc(sizeof(struct ListNode));
                new_node->val = 1;
                new_node->next = NULL;
                temp = 0;

                t1->next = new_node;
                break;
            }
            else {
                t1 = t1->next;
            }
        }
    }

    if (temp) {
        t1 = l1;
        while (t1->next)
            t1 = t1->next;

        struct ListNode* new_node = (struct ListNode*)malloc(sizeof(struct ListNode));
        new_node->val = 1;
        new_node->next = NULL;

        t1->next = new_node;
    }

    return l1;
}
